### DISCUSSION TITLE
@aws-sdk/lib-storage

### DISCUSSION THREAD

**Discussion by ppolcik-fcis** [aws/aws-sdk-js-v3#4603] [Q&A]
Is the return type of `done` function in `@aws-sdk/lib-storage` correct?

The function signature states that the return type is `Promise<AbortMultipartUploadCommandOutput | CompleteMultipartUploadCommandOutput>`, but from what I can see, the `__abortTimeout` never resolves the promises only rejects it. 

If I understand correctly, if the upload is aborted, then an error will be thrown and the`AbortMultipartUploadCommandOutput`  will not be aborted. 

Hence, I think the return type of the `done` function should be simply the `Promise<CompleteMultipartUploadCommandOutput>`.

If not, then the second part of my question is - what's the recommended way to find out, what was actually the return type after the upload has finished?

best,
Piotr

**Comment by abbasyadollahi** [comment]
Did you ever find out the solution or reasoning to this?
Running into the exact same issue and my only way to get around it as of now is to explicitly cast the output, which feels unnecessary and hacky.

```typescript
import { CompleteMultipartUploadCommandOutput, S3 } from "@aws-sdk/client-s3";
import { Upload } from "@aws-sdk/lib-storage";
...
const upload = new Upload({
  client: s3Client,
  params: s3Params,
});
const output = await upload.done() as CompleteMultipartUploadCommandOutput;
```

**Comment by mfka** [comment]
there is a more elegant solution:

```ts
import { CompleteMultipartUploadCommandOutput, S3 } from "@aws-sdk/client-s3";
import { Upload } from "@aws-sdk/lib-storage";
...
const upload = new Upload({
  client: s3Client,
  params: s3Params,
});

const data = await upload.done();

if (!("Location" in data)) {
  throw Error("Abort file upload");
}
// data is now typeof CompleteMultipartUploadCommandOutput

````
